yb-voyager/src/queryparser/traversal.go

```diff
@@ -0,0 +1,173 @@
+	We can tree each query/node as protobuf having fields and other nodes nested in it
+
+	For example:
+	Query: SELECT id, xmlelement(name "employee", name) AS employee_data FROM employees
+	ParseTree: version:160001  stmts:{stmt:{
+	select_stmt:{target_list:{res_target:{val:{column_ref:{fields:{string:{sval:"id"}}  location:7}}  location:7}}
+	target_list:{res_target:{name:"employee_data"  val:{xml_expr:{op:IS_XMLELEMENT  name:"employee"  args:{column_ref:{fields:{string:{sval:"name"}}  location:39}}  xmloption:XMLOPTION_DOCUMENT  location:11}}  location:11}}
+	from_clause:{range_var:{relname:"employees"  inh:true  relpersistence:"p"  location:67}}  limit_option:LIMIT_OPTION_DEFAULT  op:SETOP_NONE}}}
+
+	Traversal Output:
+	Traversing NodeType: pg_query.Node
+	Traversing NodeType: pg_query.SelectStmt
+	Field: target_list, Type: message
+	Traversing NodeType: pg_query.Node
+	Traversing NodeType: pg_query.ResTarget
+	Field: val, Type: message
+	Traversing NodeType: pg_query.Node
+	Traversing NodeType: pg_query.ColumnRef
+	Field: fields, Type: message
+	Traversing NodeType: pg_query.Node
+	Traversing NodeType: pg_query.String
+	Field: sval, Type: string
+	Field: location, Type: int32
+	Field: location, Type: int32
+	Traversing NodeType: pg_query.Node
+	Traversing NodeType: pg_query.ResTarget
+	Field: name, Type: string
+	Field: val, Type: message
+	Traversing NodeType: pg_query.Node
+	Traversing NodeType: pg_query.XmlExpr
+	XML Functions
+	Field: op, Type: enum
+	Field: name, Type: string
+	Field: args, Type: message
+	Traversing NodeType: pg_query.Node
+	Traversing NodeType: pg_query.ColumnRef
+	Field: fields, Type: message
+	Traversing NodeType: pg_query.Node
+	Traversing NodeType: pg_query.String
+	Field: sval, Type: string
+	Field: location, Type: int32
+	Field: xmloption, Type: enum
+	Field: location, Type: int32
+	Field: location, Type: int32
+	Field: from_clause, Type: message
+	Traversing NodeType: pg_query.Node
+	Traversing NodeType: pg_query.RangeVar
+	Field: relname, Type: string
+	Field: inh, Type: bool
+	Field: relpersistence, Type: string
+	Field: location, Type: int32
+	Field: limit_option, Type: enum
+	Field: op, Type: enum
+*/
+
+func TraverseParseTree(msg protoreflect.Message, visited map[protoreflect.Message]bool, processor NodeProcessor) error {
+	if msg == nil || !msg.IsValid() {
+		return nil
+	}
+
+	if visited[msg] {
+		return nil
+	}
+	visited[msg] = true
+
+	nodeType := msg.Descriptor().FullName()
+	log.Debugf("Traversing NodeType: %s\n", nodeType)
+	// applying the processor to the current node
+	if err := processor(msg); err != nil {
+		log.Debugf("error processing node %s: %v", nodeType, err)
+		return fmt.Errorf("error processing node %s: %w", nodeType, err)
+	}
+
+	// Reference Oneof - https://protobuf.dev/programming-guides/proto3/#oneof
+	if nodeType == PG_QUERY_NODE_NODE {
+		nodeField := msg.WhichOneof(msg.Descriptor().Oneofs().ByName("node"))
```

Comment:
call this `childNodeField` ?

Reply:
It's the same node. Just the field from OneOfs is being find out to use.